source/material_model/equation_of_state/thermodynamic_table_lookup.cc

```diff
@@ -0,0 +1,577 @@
+                                    "You have said that there is a background field and prescribed "
+                                    + Utilities::int_to_string(n_material_lookups)
+                                    + " material data files, but there are "
+                                    + Utilities::int_to_string(this->n_compositional_fields())
+                                    + " compositional fields. "
+                                    "The maximum number of material data files should therefore be "
+                                    + Utilities::int_to_string(this->n_compositional_fields() + 1)
+                                    + ". "));
+          }
+        else
+          {
+            AssertThrow ((n_material_lookups <= this->n_compositional_fields()),
+                         ExcMessage("The thermodynamic table lookup plugin requires that there is a material file "
+                                    "for every compositional field that corresponds to a mass fraction. "
+                                    "You have said that there is no background field and prescribed "
+                                    + Utilities::int_to_string(n_material_lookups)
+                                    + " material data files, but there are "
+                                    + Utilities::int_to_string(this->n_compositional_fields())
+                                    + " compositional fields. "
+                                    "The maximum number of material data files should therefore be "
+                                    + Utilities::int_to_string(this->n_compositional_fields())
+                                    + ". "));
+          }
+
+        if (latent_heat)
+          AssertThrow (n_material_lookups == 1,
+                       ExcMessage("Isochemical latent heat calculations are only implemented for a single material lookup."));
+
+      }
+
+
+
+      template <int dim>
+      void
+      ThermodynamicTableLookup<dim>::fill_additional_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
+                                                              const std::vector<std::vector<double>> &volume_fractions,
+                                                              MaterialModel::MaterialModelOutputs<dim> &out) const
+      {
+        // fill seismic velocity outputs if they exist
+        if (SeismicAdditionalOutputs<dim> *seismic_out = out.template get_additional_output<SeismicAdditionalOutputs<dim> >())
+          fill_seismic_velocities(in, out.densities, volume_fractions, seismic_out);
+
+        // fill phase volume outputs if they exist
+        if (NamedAdditionalMaterialOutputs<dim> *phase_volume_fractions_out = out.template get_additional_output<NamedAdditionalMaterialOutputs<dim> >())
+          fill_phase_volume_fractions(in, volume_fractions, phase_volume_fractions_out);
+      }
```

Comment:
Can you move that implementation above the one of the declare parameters function?